File: app/src/main/java/com/chaitupenju/popularmovies2/MovieAdapter.java
package com.chaitupenju.popularmovies2;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.chaitupenju.popularmovies2.datautils.MovieDetails;
import com.squareup.picasso.Picasso;


public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MovieAdapterViewHolder> {
    Context context;
    private MovieDetails[] mMovieData;
    public static final String BASE_IMAGE_URL1 = "https://image.tmdb.org/t/p/w500";
    public static final String BASE_IMAGE_URL2 = "?api_key=e62aae5e1c8389286f8dd8a887787ff7&page=1&language=en-US";

    private final MovieAdapterClickHandler mMovieClickHandler;

    public interface MovieAdapterClickHandler{
        void onClick(MovieDetails mIntent);
    }

    public MovieAdapter(Context context, MovieAdapterClickHandler mMovieClickHandler){
        this.context = context;
        this.mMovieClickHandler = mMovieClickHandler;
    }
    @NonNull
    @Override
    public MovieAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.movie_list_item, parent, false);
        return new MovieAdapterViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MovieAdapterViewHolder holder, int position) {
        MovieDetails movieDetail = mMovieData[position];
        Picasso.with(context)
                .load(BASE_IMAGE_URL1 + movieDetail.getImage_url() + BASE_IMAGE_URL2)
                .into(holder.mMovieImageView);
    }

    @Override
    public int getItemCount() {
        if(null == mMovieData) return 0;
        return mMovieData.length;
    }

    public class MovieAdapterViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener{
        public final ImageView mMovieImageView;
        public MovieAdapterViewHolder(View itemView) {
            super(itemView);
            mMovieImageView = itemView.findViewById(R.id.movie_img);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            int adapterPos = getAdapterPosition();
            MovieDetails movieData = mMovieData[adapterPos];
            mMovieClickHandler.onClick(movieData);
        }
    }
    public void setMovieData(MovieDetails[] movieData) {
        mMovieData = movieData;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/chaitupenju/MovieAdapter.kt
package com.chaitupenju

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.chaitupenju.popularmovies2.databinding.MovieListItemBinding
import com.chaitupenju.popularmovies2.datautils.MovieDetails
import com.chaitupenju.popularmovies2.models.MovieDetail
import com.squareup.picasso.Picasso

class MovieAdapter(
    private val context: Context,
    private val mMovieClickHandler: (MovieDetail) -> Unit
) : RecyclerView.Adapter<MovieAdapter.MovieAdapterViewHolder>() {

    var mMovieData: List<MovieDetail> = listOf()

    companion object {
        const val BASE_IMAGE_URL1 = "https://image.tmdb.org/t/p/w500"
        const val BASE_IMAGE_URL2 =
            "?api_key=e62aae5e1c8389286f8dd8a887787ff7&page=1&language=en-US"
    }

    fun setMovieData(movieData: List<MovieDetail>) {
        mMovieData = movieData
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): MovieAdapter.MovieAdapterViewHolder {
        return MovieAdapterViewHolder(
            MovieListItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        )
    }

    override fun onBindViewHolder(holder: MovieAdapter.MovieAdapterViewHolder, position: Int) {
        val movieDetail = mMovieData[position]
        holder.onBind(movieDetail = movieDetail)

        holder.itemView.setOnClickListener { mMovieClickHandler.invoke(movieDetail) }
    }

    override fun getItemCount(): Int = mMovieData.size

    inner class MovieAdapterViewHolder(private val binding: MovieListItemBinding) : RecyclerView.ViewHolder(binding.root) {

        fun onBind(movieDetail: MovieDetail) {
            Picasso.Builder(context)
                .build()
                .load(BASE_IMAGE_URL1 + movieDetail.posterPath + BASE_IMAGE_URL2)
                .into(binding.movieImg)
        }
    }
}